File: integrations/sqlalchemy/models/mysql.py
"""
Comprehensive MySQL models exercising all features dbhelper must support.

Feature coverage:
 - Column types: TINYINT, SMALLINT, INT, BIGINT (unsigned), FLOAT, DOUBLE,
   DECIMAL, VARCHAR, CHAR, TEXT, MEDIUMTEXT, LONGTEXT, BOOLEAN, DATE, TIME,
   DATETIME (fsp), TIMESTAMP (fsp), YEAR, JSON, BINARY, VARBINARY, BLOB,
   ENUM, SET
 - Primary keys: single-column, composite
 - Foreign keys: simple, composite, cascade/set-null/restrict
 - Indexes: btree, unique, composite
 - Constraints: not-null, unique, check, default
 - Auto-increment
 - ON UPDATE CURRENT_TIMESTAMP
"""

import enum

from sqlalchemy import (
    BigInteger,
    Boolean,
    CheckConstraint,
    Date,
    DateTime,
    Float,
    ForeignKey,
    ForeignKeyConstraint,
    Index,
    Integer,
    Numeric,
    SmallInteger,
    String,
    Text,
    UniqueConstraint,
    text,
)
from sqlalchemy.dialects.mysql import (
    BIGINT,
    BINARY,
    CHAR,
    DATETIME as MYSQL_DATETIME,
    DOUBLE,
    ENUM,
    INTEGER,
    JSON,
    LONGTEXT,
    MEDIUMTEXT,
    SET,
    SMALLINT,
    TIMESTAMP as MYSQL_TIMESTAMP,
    TINYINT,
    VARBINARY,
    VARCHAR,
    YEAR,
)
from sqlalchemy.orm import Mapped, mapped_column

from models.base import Base


# ---------------------------------------------------------------------------
# Tables
# ---------------------------------------------------------------------------


class MyUser(Base):
    """Core users table for MySQL."""

    __tablename__ = "my_users"
    __table_args__ = (
        UniqueConstraint("uuid", name="my_users_uuid_uniq"),
        Index("my_users_email_idx", "email", unique=True),
        Index("my_users_username_idx", "username", unique=True),
        Index("my_users_role_idx", "role"),
        Index("my_users_created_at_idx", "created_at"),
        CheckConstraint(
            "age IS NULL OR (age >= 0 AND age <= 200)", name="my_users_age_check"
        ),
        {"mysql_engine": "InnoDB", "mysql_charset": "utf8mb4"},
    )

    id: Mapped[int] = mapped_column(
        BIGINT(unsigned=True), primary_key=True, autoincrement=True
    )
    uuid: Mapped[str] = mapped_column(
        VARCHAR(36), nullable=False, server_default=text("(UUID())")
    )
    email: Mapped[str] = mapped_column(VARCHAR(255), nullable=False)
    username: Mapped[str] = mapped_column(VARCHAR(100), nullable=False)
    password_hash: Mapped[str] = mapped_column(Text, nullable=False)
    role: Mapped[str] = mapped_column(
        ENUM("admin", "editor", "viewer", "guest"),
        nullable=False,
        server_default="viewer",
    )
    display_name: Mapped[str | None] = mapped_column(VARCHAR(200))
    bio: Mapped[str | None] = mapped_column(MEDIUMTEXT)
    age: Mapped[int | None] = mapped_column(TINYINT(unsigned=True))
    score: Mapped[float | None] = mapped_column(Float, server_default=text("0"))
    balance: Mapped[str | None] = mapped_column(
        Numeric(15, 2), server_default=text("0.00")
    )
    is_active: Mapped[bool] = mapped_column(
        Boolean, nullable=False, server_default=text("1")
    )
    metadata_: Mapped[str | None] = mapped_column("metadata", JSON)
    last_login_at: Mapped[str | None] = mapped_column(MYSQL_DATETIME(fsp=3))
    created_at: Mapped[str] = mapped_column(
        MYSQL_TIMESTAMP(fsp=3),
        nullable=False,
        server_default=text("CURRENT_TIMESTAMP(3)"),
    )
    updated_at: Mapped[str] = mapped_column(
        MYSQL_TIMESTAMP(fsp=3),
        nullable=False,
        server_default=text("CURRENT_TIMESTAMP(3) ON UPDATE CURRENT_TIMESTAMP(3)"),
    )


class MyOrganization(Base):
    """Organizations for MySQL."""

    __tablename__ = "my_organizations"
    __table_args__ = (
        UniqueConstraint("slug"),
        {"mysql_engine": "InnoDB", "mysql_charset": "utf8mb4"},
    )

    id: Mapped[int] = mapped_column(
        INTEGER(unsigned=True), primary_key=True, autoincrement=True
    )
    name: Mapped[str] = mapped_column(VARCHAR(255), nullable=False)
    slug: Mapped[str] = mapped_column(VARCHAR(100), nullable=False)
    parent_id: Mapped[int | None] = mapped_column(
        INTEGER(unsigned=True),
        ForeignKey("my_organizations.id", ondelete="SET NULL"),
    )
    description: Mapped[str | None] = mapped_column(Text)
    website: Mapped[str | None] = mapped_column(VARCHAR(500))
    logo_url: Mapped[str | None] = mapped_column(Text)
    is_verified: Mapped[bool] = mapped_column(
        Boolean, nullable=False, server_default=text("0")
    )
    settings: Mapped[str | None] = mapped_column(JSON)
    created_at: Mapped[str] = mapped_column(
        MYSQL_TIMESTAMP(fsp=3),
        nullable=False,
        server_default=text("CURRENT_TIMESTAMP(3)"),
    )


class MyOrgMember(Base):
    """Org members — composite PK for MySQL."""

    __tablename__ = "my_org_members"
    __table_args__ = (
        ForeignKeyConstraint(
            ["org_id"],
            ["my_organizations.id"],
            ondelete="CASCADE",
            onupdate="CASCADE",
        ),
        ForeignKeyConstraint(
            ["user_id"],
            ["my_users.id"],
            ondelete="CASCADE",
            onupdate="CASCADE",
        ),
        {"mysql_engine": "InnoDB", "mysql_charset": "utf8mb4"},
    )

    org_id: Mapped[int] = mapped_column(INTEGER(unsigned=True), primary_key=True)
    user_id: Mapped[int] = mapped_column(BIGINT(unsigned=True), primary_key=True)
    role: Mapped[str] = mapped_column(
        VARCHAR(50), nullable=False, server_default=text("'member'")
    )
    joined_at: Mapped[str] = mapped_column(
        MYSQL_TIMESTAMP(fsp=3),
        nullable=False,
        server_default=text("CURRENT_TIMESTAMP(3)"),
    )


class MyProject(Base):
    """Projects for MySQL."""

    __tablename__ = "my_projects"
    __table_args__ = (
        ForeignKeyConstraint(
            ["org_id"], ["my_organizations.id"], ondelete="CASCADE"
        ),
        UniqueConstraint("org_id", "slug", name="my_projects_org_slug_uniq"),
        Index("my_projects_org_id_idx", "org_id"),
        Index("my_projects_priority_idx", "priority"),
        CheckConstraint(
            "start_date IS NULL OR deadline IS NULL OR start_date <= deadline",
            name="my_projects_dates_check",
        ),
        {"mysql_engine": "InnoDB", "mysql_charset": "utf8mb4"},
    )

    id: Mapped[int] = mapped_column(
        INTEGER(unsigned=True), primary_key=True, autoincrement=True
    )
    org_id: Mapped[int] = mapped_column(INTEGER(unsigned=True), nullable=False)
    name: Mapped[str] = mapped_column(VARCHAR(255), nullable=False)
    slug: Mapped[str] = mapped_column(VARCHAR(100), nullable=False)
    description: Mapped[str | None] = mapped_column(Text)
    priority: Mapped[str] = mapped_column(
        ENUM("low", "medium", "high", "urgent"),
        nullable=False,
        server_default="medium",
    )
    is_archived: Mapped[bool] = mapped_column(
        Boolean, nullable=False, server_default=text("0")
    )
    config: Mapped[str | None] = mapped_column(JSON)
    start_date: Mapped[str | None] = mapped_column(Date)
    deadline: Mapped[str | None] = mapped_column(Date)
    created_at: Mapped[str] = mapped_column(
        MYSQL_TIMESTAMP(fsp=3),
        nullable=False,
        server_default=text("CURRENT_TIMESTAMP(3)"),
    )
    updated_at: Mapped[str] = mapped_column(
        MYSQL_TIMESTAMP(fsp=3),
        nullable=False,
        server_default=text("CURRENT_TIMESTAMP(3) ON UPDATE CURRENT_TIMESTAMP(3)"),
    )


class MyTask(Base):
    """Tasks for MySQL."""

    __tablename__ = "my_tasks"
    __table_args__ = (
        ForeignKeyConstraint(
            ["project_id"], ["my_projects.id"], ondelete="CASCADE"
        ),
        ForeignKeyConstraint(
            ["assignee_id"], ["my_users.id"], ondelete="SET NULL"
        ),
        Index("my_tasks_project_status_idx", "project_id", "status"),
        Index("my_tasks_assignee_idx", "assignee_id"),
        {"mysql_engine": "InnoDB", "mysql_charset": "utf8mb4"},
    )

    id: Mapped[int] = mapped_column(
        BIGINT(unsigned=True), primary_key=True, autoincrement=True
    )
    project_id: Mapped[int] = mapped_column(INTEGER(unsigned=True), nullable=False)
    parent_id: Mapped[int | None] = mapped_column(
        BIGINT(unsigned=True),
        ForeignKey("my_tasks.id", ondelete="CASCADE"),
    )
    assignee_id: Mapped[int | None] = mapped_column(BIGINT(unsigned=True))
    title: Mapped[str] = mapped_column(VARCHAR(500), nullable=False)
    description: Mapped[str | None] = mapped_column(MEDIUMTEXT)
    status: Mapped[str] = mapped_column(
        ENUM("pending", "processing", "shipped", "delivered", "cancelled", "refunded"),
        nullable=False,
        server_default="pending",
    )
    priority: Mapped[str] = mapped_column(
        ENUM("low", "medium", "high", "urgent"),
        nullable=False,
        server_default="medium",
    )
    sort_order: Mapped[int] = mapped_column(
        Integer, nullable=False, server_default=text("0")
    )
    estimate_hours: Mapped[str | None] = mapped_column(Numeric(6, 2))
    due_date: Mapped[str | None] = mapped_column(MYSQL_DATETIME(fsp=3))
    completed_at: Mapped[str | None] = mapped_column(MYSQL_DATETIME(fsp=3))
    created_at: Mapped[str] = mapped_column(
        MYSQL_TIMESTAMP(fsp=3),
        nullable=False,
        server_default=text("CURRENT_TIMESTAMP(3)"),
    )
    updated_at: Mapped[str] = mapped_column(
        MYSQL_TIMESTAMP(fsp=3),
        nullable=False,
        server_default=text("CURRENT_TIMESTAMP(3) ON UPDATE CURRENT_TIMESTAMP(3)"),
    )


class MyComment(Base):
    """Comments for MySQL."""

    __tablename__ = "my_comments"
    __table_args__ = (
        ForeignKeyConstraint(
            ["task_id"], ["my_tasks.id"], ondelete="CASCADE"
        ),
        ForeignKeyConstraint(
            ["author_id"], ["my_users.id"], ondelete="CASCADE"
        ),
        Index("my_comments_task_idx", "task_id"),
        Index("my_comments_author_idx", "author_id"),
        {"mysql_engine": "InnoDB", "mysql_charset": "utf8mb4"},
    )

    id: Mapped[int] = mapped_column(
        BIGINT(unsigned=True), primary_key=True, autoincrement=True
    )
    task_id: Mapped[int] = mapped_column(BIGINT(unsigned=True), nullable=False)
    author_id: Mapped[int] = mapped_column(BIGINT(unsigned=True), nullable=False)
    parent_comment_id: Mapped[int | None] = mapped_column(
        BIGINT(unsigned=True),
        ForeignKey("my_comments.id", ondelete="CASCADE"),
    )
    body: Mapped[str] = mapped_column(MEDIUMTEXT, nullable=False)
    edited_at: Mapped[str | None] = mapped_column(MYSQL_DATETIME(fsp=3))
    created_at: Mapped[str] = mapped_column(
        MYSQL_TIMESTAMP(fsp=3),
        nullable=False,
        server_default=text("CURRENT_TIMESTAMP(3)"),
    )


class MyTag(Base):
    """Tags for MySQL."""

    __tablename__ = "my_tags"
    __table_args__ = (
        UniqueConstraint("name"),
        {"mysql_engine": "InnoDB", "mysql_charset": "utf8mb4"},
    )

    id: Mapped[int] = mapped_column(
        INTEGER(unsigned=True), primary_key=True, autoincrement=True
    )
    name: Mapped[str] = mapped_column(VARCHAR(100), nullable=False)
    color: Mapped[str] = mapped_column(
        CHAR(7), nullable=False, server_default=text("'#808080'")
    )
    description: Mapped[str | None] = mapped_column(Text)
    created_at: Mapped[str] = mapped_column(
        MYSQL_TIMESTAMP(fsp=3),
        nullable=False,
        server_default=text("CURRENT_TIMESTAMP(3)"),
    )


class MyTaskTag(Base):
    """Join table: tasks <-> tags for MySQL."""

    __tablename__ = "my_task_tags"
    __table_args__ = (
        ForeignKeyConstraint(
            ["task_id"], ["my_tasks.id"], ondelete="CASCADE"
        ),
        ForeignKeyConstraint(
            ["tag_id"], ["my_tags.id"], ondelete="CASCADE"
        ),
        {"mysql_engine": "InnoDB", "mysql_charset": "utf8mb4"},
    )

    task_id: Mapped[int] = mapped_column(BIGINT(unsigned=True), primary_key=True)
    tag_id: Mapped[int] = mapped_column(INTEGER(unsigned=True), primary_key=True)
    added_at: Mapped[str] = mapped_column(
        MYSQL_TIMESTAMP(fsp=3),
        nullable=False,
        server_default=text("CURRENT_TIMESTAMP(3)"),
    )


class MyAuditLog(Base):
    """Audit log for MySQL."""

    __tablename__ = "my_audit_log"
    __table_args__ = (
        ForeignKeyConstraint(
            ["user_id"], ["my_users.id"], ondelete="SET NULL"
        ),
        Index("my_audit_log_user_idx", "user_id"),
        Index("my_audit_log_table_action_idx", "table_name", "action"),
        Index("my_audit_log_occurred_at_idx", "occurred_at"),
        {"mysql_engine": "InnoDB", "mysql_charset": "utf8mb4"},
    )

    id: Mapped[int] = mapped_column(
        BIGINT(unsigned=True), primary_key=True, autoincrement=True
    )
    user_id: Mapped[int | None] = mapped_column(BIGINT(unsigned=True))
    action: Mapped[str] = mapped_column(VARCHAR(100), nullable=False)
    table_name: Mapped[str] = mapped_column(VARCHAR(100), nullable=False)
    record_id: Mapped[str | None] = mapped_column(VARCHAR(255))
    old_values: Mapped[str | None] = mapped_column(JSON)
    new_values: Mapped[str | None] = mapped_column(JSON)
    ip_address: Mapped[str | None] = mapped_column(VARCHAR(45))
    user_agent: Mapped[str | None] = mapped_column(Text)
    occurred_at: Mapped[str] = mapped_column(
        MYSQL_TIMESTAMP(fsp=3),
        nullable=False,
        server_default=text("CURRENT_TIMESTAMP(3)"),
    )


class MyAttachment(Base):
    """Attachments for MySQL."""

    __tablename__ = "my_attachments"
    __table_args__ = (
        ForeignKeyConstraint(
            ["task_id"], ["my_tasks.id"], ondelete="CASCADE"
        ),
        ForeignKeyConstraint(
            ["uploaded_by"], ["my_users.id"], ondelete="CASCADE"
        ),
        Index("my_attachments_task_idx", "task_id"),
        CheckConstraint("size_bytes > 0", name="my_attachments_size_check"),
        {"mysql_engine": "InnoDB", "mysql_charset": "utf8mb4"},
    )

    id: Mapped[str] = mapped_column(
        VARCHAR(36), primary_key=True, server_default=text("(UUID())")
    )
    task_id: Mapped[int] = mapped_column(BIGINT(unsigned=True), nullable=False)
    uploaded_by: Mapped[int] = mapped_column(BIGINT(unsigned=True), nullable=False)
    file_name: Mapped[str] = mapped_column(VARCHAR(500), nullable=False)
    mime_type: Mapped[str] = mapped_column(VARCHAR(255), nullable=False)
    size_bytes: Mapped[int] = mapped_column(BIGINT(unsigned=True), nullable=False)
    storage_path: Mapped[str] = mapped_column(Text, nullable=False)
    checksum: Mapped[str | None] = mapped_column(VARCHAR(128))
    metadata_: Mapped[str | None] = mapped_column("metadata", JSON)
    created_at: Mapped[str] = mapped_column(
        MYSQL_TIMESTAMP(fsp=3),
        nullable=False,
        server_default=text("CURRENT_TIMESTAMP(3)"),
    )


class MyNotification(Base):
    """Notifications for MySQL."""

    __tablename__ = "my_notifications"
    __table_args__ = (
        ForeignKeyConstraint(
            ["user_id"], ["my_users.id"], ondelete="CASCADE"
        ),
        Index("my_notifications_user_idx", "user_id"),
        Index("my_notifications_created_at_idx", "created_at"),
        {"mysql_engine": "InnoDB", "mysql_charset": "utf8mb4"},
    )

    id: Mapped[int] = mapped_column(
        BIGINT(unsigned=True), primary_key=True, autoincrement=True
    )
    user_id: Mapped[int] = mapped_column(BIGINT(unsigned=True), nullable=False)
    type: Mapped[str] = mapped_column(VARCHAR(50), nullable=False)
    title: Mapped[str] = mapped_column(VARCHAR(255), nullable=False)
    body: Mapped[str | None] = mapped_column(Text)
    data: Mapped[str | None] = mapped_column(JSON)
    is_read: Mapped[bool] = mapped_column(
        Boolean, nullable=False, server_default=text("0")
    )
    read_at: Mapped[str | None] = mapped_column(MYSQL_DATETIME(fsp=3))
    created_at: Mapped[str] = mapped_column(
        MYSQL_TIMESTAMP(fsp=3),
        nullable=False,
        server_default=text("CURRENT_TIMESTAMP(3)"),
    )


class MySession(Base):
    """Sessions — exercises binary columns."""

    __tablename__ = "my_sessions"
    __table_args__ = (
        ForeignKeyConstraint(
            ["user_id"], ["my_users.id"], ondelete="CASCADE"
        ),
        Index("my_sessions_user_idx", "user_id"),
        Index("my_sessions_expires_idx", "expires_at"),
        {"mysql_engine": "InnoDB", "mysql_charset": "utf8mb4"},
    )

    id: Mapped[str] = mapped_column(VARCHAR(128), primary_key=True)
    user_id: Mapped[int] = mapped_column(BIGINT(unsigned=True), nullable=False)
    token_hash: Mapped[bytes] = mapped_column(BINARY(32), nullable=False)
    user_agent: Mapped[str | None] = mapped_column(Text)
    ip_address: Mapped[str | None] = mapped_column(VARCHAR(45))
    last_active_at: Mapped[str] = mapped_column(MYSQL_DATETIME(fsp=3), nullable=False)
    expires_at: Mapped[str] = mapped_column(MYSQL_DATETIME(fsp=3), nullable=False)
    created_at: Mapped[str] = mapped_column(
        MYSQL_TIMESTAMP(fsp=3),
        nullable=False,
        server_default=text("CURRENT_TIMESTAMP(3)"),
    )


class MyProjectInvite(Base):
    """Project invites — composite FK for MySQL."""

    __tablename__ = "my_project_invites"
    __table_args__ = (
        ForeignKeyConstraint(
            ["project_id"], ["my_projects.id"], ondelete="CASCADE"
        ),
        ForeignKeyConstraint(
            ["inviter_org_id", "inviter_user_id"],
            ["my_org_members.org_id", "my_org_members.user_id"],
            ondelete="CASCADE",
        ),
        Index("my_project_invites_project_idx", "project_id"),
        Index("my_project_invites_email_idx", "invitee_email"),
        {"mysql_engine": "InnoDB", "mysql_charset": "utf8mb4"},
    )

    id: Mapped[int] = mapped_column(
        INTEGER(unsigned=True), primary_key=True, autoincrement=True
    )
    project_id: Mapped[int] = mapped_column(INTEGER(unsigned=True), nullable=False)
    inviter_org_id: Mapped[int] = mapped_column(INTEGER(unsigned=True), nullable=False)
    inviter_user_id: Mapped[int] = mapped_column(
        BIGINT(unsigned=True), nullable=False
    )
    invitee_email: Mapped[str] = mapped_column(VARCHAR(255), nullable=False)
    token: Mapped[str] = mapped_column(
        VARCHAR(36), nullable=False, unique=True, server_default=text("(UUID())")
    )
    expires_at: Mapped[str] = mapped_column(MYSQL_DATETIME(fsp=3), nullable=False)
    accepted_at: Mapped[str | None] = mapped_column(MYSQL_DATETIME(fsp=3))
    created_at: Mapped[str] = mapped_column(
        MYSQL_TIMESTAMP(fsp=3),
        nullable=False,
        server_default=text("CURRENT_TIMESTAMP(3)"),
    )

File: integrations/drizzle/src/schema/mysql.ts
/**
 * Comprehensive MySQL schema exercising all features dbhelper must support.
 *
 * Feature coverage:
 *  - Column types: tinyint, smallint, int, bigint, float, double, decimal,
 *    varchar, char, text, mediumtext, longtext, boolean, date, time, datetime,
 *    timestamp, year, json, binary, varbinary, blob, enum, set
 *  - Primary keys: single-column, composite
 *  - Foreign keys: simple, composite, cascade/set-null/restrict actions
 *  - Indexes: btree, unique, composite, prefix-length
 *  - Constraints: not-null, unique, check, default (static & expression)
 *  - MySQL enums (inline) and sets
 *  - Auto-increment
 *  - Character sets and collations (via column options where possible)
 */

import {
  mysqlTable,
  mysqlEnum,
  mysqlSchema,
  serial,
  tinyint,
  smallint,
  int,
  bigint,
  float,
  double,
  decimal,
  varchar,
  char,
  text,
  mediumtext,
  longtext,
  boolean,
  date,
  time,
  datetime,
  timestamp,
  year,
  json,
  binary,
  varbinary,
  primaryKey,
  foreignKey,
  unique,
  uniqueIndex,
  index,
  check,
} from "drizzle-orm/mysql-core";
import { sql } from "drizzle-orm";

// ---------------------------------------------------------------------------
// Tables
// ---------------------------------------------------------------------------

/** Core users table */
export const users = mysqlTable(
  "users",
  {
    id: bigint("id", { mode: "number", unsigned: true })
      .primaryKey()
      .autoincrement(),
    uuid: varchar("uuid", { length: 36 })
      .notNull()
      .default(sql`(UUID())`),
    email: varchar("email", { length: 255 }).notNull(),
    username: varchar("username", { length: 100 }).notNull(),
    passwordHash: text("password_hash").notNull(),
    role: mysqlEnum("role", ["admin", "editor", "viewer", "guest"])
      .notNull()
      .default("viewer"),
    displayName: varchar("display_name", { length: 200 }),
    bio: mediumtext("bio"),
    age: tinyint("age", { unsigned: true }),
    score: float("score").default(0),
    balance: decimal("balance", { precision: 15, scale: 2 }).default("0.00"),
    isActive: boolean("is_active").notNull().default(true),
    metadata: json("metadata"),
    lastLoginAt: datetime("last_login_at", { fsp: 3 }),
    createdAt: timestamp("created_at", { fsp: 3 })
      .notNull()
      .defaultNow(),
    updatedAt: timestamp("updated_at", { fsp: 3 })
      .notNull()
      .defaultNow()
      .onUpdateNow(),
  },
  (t) => [
    uniqueIndex("users_email_idx").on(t.email),
    uniqueIndex("users_username_idx").on(t.username),
    unique("users_uuid_uniq").on(t.uuid),
    index("users_role_idx").on(t.role),
    index("users_created_at_idx").on(t.createdAt),
    check("users_age_check", sql`age IS NULL OR (age >= 0 AND age <= 200)`),
  ]
);

/** Organizations */
export const organizations = mysqlTable("organizations", {
  id: int("id", { unsigned: true }).primaryKey().autoincrement(),
  name: varchar("name", { length: 255 }).notNull(),
  slug: varchar("slug", { length: 100 }).notNull().unique(),
  parentId: int("parent_id", { unsigned: true }),
  description: text("description"),
  website: varchar("website", { length: 500 }),
  logoUrl: text("logo_url"),
  isVerified: boolean("is_verified").notNull().default(false),
  settings: json("settings"),
  createdAt: timestamp("created_at", { fsp: 3 }).notNull().defaultNow(),
});

/** Org members — composite PK */
export const orgMembers = mysqlTable(
  "org_members",
  {
    orgId: int("org_id", { unsigned: true }).notNull(),
    userId: bigint("user_id", { mode: "number", unsigned: true }).notNull(),
    role: varchar("role", { length: 50 }).notNull().default("member"),
    joinedAt: timestamp("joined_at", { fsp: 3 }).notNull().defaultNow(),
  },
  (t) => [
    primaryKey({ columns: [t.orgId, t.userId] }),
    foreignKey({ columns: [t.orgId], foreignColumns: [organizations.id] })
      .onDelete("cascade")
      .onUpdate("cascade"),
    foreignKey({ columns: [t.userId], foreignColumns: [users.id] })
      .onDelete("cascade")
      .onUpdate("cascade"),
  ]
);

/** Projects */
export const projects = mysqlTable(
  "projects",
  {
    id: int("id", { unsigned: true }).primaryKey().autoincrement(),
    orgId: int("org_id", { unsigned: true }).notNull(),
    name: varchar("name", { length: 255 }).notNull(),
    slug: varchar("slug", { length: 100 }).notNull(),
    description: text("description"),
    priority: mysqlEnum("priority", ["low", "medium", "high", "urgent"])
      .notNull()
      .default("medium"),
    isArchived: boolean("is_archived").notNull().default(false),
    config: json("config"),
    startDate: date("start_date"),
    deadline: date("deadline"),
    createdAt: timestamp("created_at", { fsp: 3 }).notNull().defaultNow(),
    updatedAt: timestamp("updated_at", { fsp: 3 })
      .notNull()
      .defaultNow()
      .onUpdateNow(),
  },
  (t) => [
    foreignKey({ columns: [t.orgId], foreignColumns: [organizations.id] })
      .onDelete("cascade"),
    unique("projects_org_slug_uniq").on(t.orgId, t.slug),
    index("projects_org_id_idx").on(t.orgId),
    index("projects_priority_idx").on(t.priority),
    check(
      "projects_dates_check",
      sql`start_date IS NULL OR deadline IS NULL OR start_date <= deadline`
    ),
  ]
);

/** Tasks */
export const tasks = mysqlTable(
  "tasks",
  {
    id: bigint("id", { mode: "number", unsigned: true })
      .primaryKey()
      .autoincrement(),
    projectId: int("project_id", { unsigned: true }).notNull(),
    parentId: bigint("parent_id", { mode: "number", unsigned: true }),
    assigneeId: bigint("assignee_id", { mode: "number", unsigned: true }),
    title: varchar("title", { length: 500 }).notNull(),
    description: mediumtext("description"),
    status: mysqlEnum("status", [
      "pending",
      "processing",
      "shipped",
      "delivered",
      "cancelled",
      "refunded",
    ])
      .notNull()
      .default("pending"),
    priority: mysqlEnum("priority", ["low", "medium", "high", "urgent"])
      .notNull()
      .default("medium"),
    sortOrder: int("sort_order").notNull().default(0),
    estimateHours: decimal("estimate_hours", { precision: 6, scale: 2 }),
    dueDate: datetime("due_date", { fsp: 3 }),
    completedAt: datetime("completed_at", { fsp: 3 }),
    createdAt: timestamp("created_at", { fsp: 3 }).notNull().defaultNow(),
    updatedAt: timestamp("updated_at", { fsp: 3 })
      .notNull()
      .defaultNow()
      .onUpdateNow(),
  },
  (t) => [
    foreignKey({ columns: [t.projectId], foreignColumns: [projects.id] })
      .onDelete("cascade"),
    foreignKey({ columns: [t.assigneeId], foreignColumns: [users.id] })
      .onDelete("set null"),
    index("tasks_project_status_idx").on(t.projectId, t.status),
    index("tasks_assignee_idx").on(t.assigneeId),
  ]
);

/** Comments */
export const comments = mysqlTable(
  "comments",
  {
    id: bigint("id", { mode: "number", unsigned: true })
      .primaryKey()
      .autoincrement(),
    taskId: bigint("task_id", { mode: "number", unsigned: true }).notNull(),
    authorId: bigint("author_id", { mode: "number", unsigned: true }).notNull(),
    parentCommentId: bigint("parent_comment_id", {
      mode: "number",
      unsigned: true,
    }),
    body: mediumtext("body").notNull(),
    editedAt: datetime("edited_at", { fsp: 3 }),
    createdAt: timestamp("created_at", { fsp: 3 }).notNull().defaultNow(),
  },
  (t) => [
    foreignKey({ columns: [t.taskId], foreignColumns: [tasks.id] })
      .onDelete("cascade"),
    foreignKey({ columns: [t.authorId], foreignColumns: [users.id] })
      .onDelete("cascade"),
    index("comments_task_idx").on(t.taskId),
    index("comments_author_idx").on(t.authorId),
  ]
);

/** Tags */
export const tags = mysqlTable("tags", {
  id: int("id", { unsigned: true }).primaryKey().autoincrement(),
  name: varchar("name", { length: 100 }).notNull().unique(),
  color: char("color", { length: 7 }).notNull().default("#808080"),
  description: text("description"),
  createdAt: timestamp("created_at", { fsp: 3 }).notNull().defaultNow(),
});

/** Join table: tasks <-> tags */
export const taskTags = mysqlTable(
  "task_tags",
  {
    taskId: bigint("task_id", { mode: "number", unsigned: true }).notNull(),
    tagId: int("tag_id", { unsigned: true }).notNull(),
    addedAt: timestamp("added_at", { fsp: 3 }).notNull().defaultNow(),
  },
  (t) => [
    primaryKey({ columns: [t.taskId, t.tagId] }),
    foreignKey({ columns: [t.taskId], foreignColumns: [tasks.id] })
      .onDelete("cascade"),
    foreignKey({ columns: [t.tagId], foreignColumns: [tags.id] })
      .onDelete("cascade"),
  ]
);

/** Audit log */
export const auditLog = mysqlTable(
  "audit_log",
  {
    id: bigint("id", { mode: "number", unsigned: true })
      .primaryKey()
      .autoincrement(),
    userId: bigint("user_id", { mode: "number", unsigned: true }),
    action: varchar("action", { length: 100 }).notNull(),
    tableName: varchar("table_name", { length: 100 }).notNull(),
    recordId: varchar("record_id", { length: 255 }),
    oldValues: json("old_values"),
    newValues: json("new_values"),
    ipAddress: varchar("ip_address", { length: 45 }),
    userAgent: text("user_agent"),
    occurredAt: timestamp("occurred_at", { fsp: 3 }).notNull().defaultNow(),
  },
  (t) => [
    foreignKey({ columns: [t.userId], foreignColumns: [users.id] })
      .onDelete("set null"),
    index("audit_log_user_idx").on(t.userId),
    index("audit_log_table_action_idx").on(t.tableName, t.action),
    index("audit_log_occurred_at_idx").on(t.occurredAt),
  ]
);

/** Attachments */
export const attachments = mysqlTable(
  "attachments",
  {
    id: varchar("id", { length: 36 })
      .primaryKey()
      .default(sql`(UUID())`),
    taskId: bigint("task_id", { mode: "number", unsigned: true }).notNull(),
    uploadedBy: bigint("uploaded_by", { mode: "number", unsigned: true })
      .notNull(),
    fileName: varchar("file_name", { length: 500 }).notNull(),
    mimeType: varchar("mime_type", { length: 255 }).notNull(),
    sizeBytes: bigint("size_bytes", { mode: "number", unsigned: true })
      .notNull(),
    storagePath: text("storage_path").notNull(),
    checksum: varchar("checksum", { length: 128 }),
    metadata: json("metadata"),
    createdAt: timestamp("created_at", { fsp: 3 }).notNull().defaultNow(),
  },
  (t) => [
    foreignKey({ columns: [t.taskId], foreignColumns: [tasks.id] })
      .onDelete("cascade"),
    foreignKey({ columns: [t.uploadedBy], foreignColumns: [users.id] })
      .onDelete("cascade"),
    index("attachments_task_idx").on(t.taskId),
    check("attachments_size_check", sql`size_bytes > 0`),
  ]
);

/** Notifications */
export const notifications = mysqlTable(
  "notifications",
  {
    id: bigint("id", { mode: "number", unsigned: true })
      .primaryKey()
      .autoincrement(),
    userId: bigint("user_id", { mode: "number", unsigned: true }).notNull(),
    type: varchar("type", { length: 50 }).notNull(),
    title: varchar("title", { length: 255 }).notNull(),
    body: text("body"),
    data: json("data"),
    isRead: boolean("is_read").notNull().default(false),
    readAt: datetime("read_at", { fsp: 3 }),
    createdAt: timestamp("created_at", { fsp: 3 }).notNull().defaultNow(),
  },
  (t) => [
    foreignKey({ columns: [t.userId], foreignColumns: [users.id] })
      .onDelete("cascade"),
    index("notifications_user_idx").on(t.userId),
    index("notifications_created_at_idx").on(t.createdAt),
  ]
);

/** Sessions — exercises binary columns and TTL-like patterns */
export const sessions = mysqlTable(
  "sessions",
  {
    id: varchar("id", { length: 128 }).primaryKey(),
    userId: bigint("user_id", { mode: "number", unsigned: true }).notNull(),
    tokenHash: binary("token_hash", { length: 32 }).notNull(),
    userAgent: text("user_agent"),
    ipAddress: varchar("ip_address", { length: 45 }),
    lastActiveAt: datetime("last_active_at", { fsp: 3 }).notNull(),
    expiresAt: datetime("expires_at", { fsp: 3 }).notNull(),
    createdAt: timestamp("created_at", { fsp: 3 }).notNull().defaultNow(),
  },
  (t) => [
    foreignKey({ columns: [t.userId], foreignColumns: [users.id] })
      .onDelete("cascade"),
    index("sessions_user_idx").on(t.userId),
    index("sessions_expires_idx").on(t.expiresAt),
  ]
);

/** Composite FK example — project invites */
export const projectInvites = mysqlTable(
  "project_invites",
  {
    id: int("id", { unsigned: true }).primaryKey().autoincrement(),
    projectId: int("project_id", { unsigned: true }).notNull(),
    inviterOrgId: int("inviter_org_id", { unsigned: true }).notNull(),
    inviterUserId: bigint("inviter_user_id", {
      mode: "number",
      unsigned: true,
    }).notNull(),
    inviteeEmail: varchar("invitee_email", { length: 255 }).notNull(),
    token: varchar("token", { length: 36 })
      .notNull()
      .unique()
      .default(sql`(UUID())`),
    expiresAt: datetime("expires_at", { fsp: 3 }).notNull(),
    acceptedAt: datetime("accepted_at", { fsp: 3 }),
    createdAt: timestamp("created_at", { fsp: 3 }).notNull().defaultNow(),
  },
  (t) => [
    foreignKey({ columns: [t.projectId], foreignColumns: [projects.id] })
      .onDelete("cascade"),
    foreignKey({
      columns: [t.inviterOrgId, t.inviterUserId],
      foreignColumns: [orgMembers.orgId, orgMembers.userId],
    }).onDelete("cascade"),
    index("project_invites_project_idx").on(t.projectId),
    index("project_invites_email_idx").on(t.inviteeEmail),
  ]
);
